src/main/java/jcifs/dcerpc/msrpc/lsarpc.java

```diff
@@ -0,0 +1,1046 @@
+                _dst.enc_ndr_small(this.domain_guid.node[ _i ]);
+            }
+            if ( this.sid != null ) {
+                _dst = _dst.deferred;
+                this.sid.encode(_dst);
+
+            }
+        }
+
+
+        @Override
+        public void decode ( NdrBuffer _src ) throws NdrException {
+            _src.align(4);
+            _src.align(4);
+            if ( this.name == null ) {
+                this.name = new rpc.unicode_string();
+            }
+            this.name.length = (short) _src.dec_ndr_short();
+            this.name.maximum_length = (short) _src.dec_ndr_short();
+            int _name_bufferp = _src.dec_ndr_long();
+            _src.align(4);
+            if ( this.dns_domain == null ) {
+                this.dns_domain = new rpc.unicode_string();
+            }
+            this.dns_domain.length = (short) _src.dec_ndr_short();
+            this.dns_domain.maximum_length = (short) _src.dec_ndr_short();
+            int _dns_domain_bufferp = _src.dec_ndr_long();
+            _src.align(4);
+            if ( this.dns_forest == null ) {
+                this.dns_forest = new rpc.unicode_string();
+            }
+            this.dns_forest.length = (short) _src.dec_ndr_short();
+            this.dns_forest.maximum_length = (short) _src.dec_ndr_short();
+            int _dns_forest_bufferp = _src.dec_ndr_long();
+            _src.align(4);
+            if ( this.domain_guid == null ) {
+                this.domain_guid = new rpc.uuid_t();
+            }
+            this.domain_guid.time_low = _src.dec_ndr_long();
+            this.domain_guid.time_mid = (short) _src.dec_ndr_short();
+            this.domain_guid.time_hi_and_version = (short) _src.dec_ndr_short();
+            this.domain_guid.clock_seq_hi_and_reserved = (byte) _src.dec_ndr_small();
+            this.domain_guid.clock_seq_low = (byte) _src.dec_ndr_small();
+            int _domain_guid_nodes = 6;
+            int _domain_guid_nodei = _src.index;
+            _src.advance(1 * _domain_guid_nodes);
+            int _sidp = _src.dec_ndr_long();
+
+            if ( _name_bufferp != 0 ) {
+                _src = _src.deferred;
+                int _name_buffers = _src.dec_ndr_long();
+                _src.dec_ndr_long();
+                int _name_bufferl = _src.dec_ndr_long();
+                int _name_bufferi = _src.index;
+                _src.advance(2 * _name_bufferl);
+
+                if ( this.name.buffer == null ) {
+                    if ( _name_buffers < 0 || _name_buffers > 0xFFFF )
+                        throw new NdrException(NdrException.INVALID_CONFORMANCE);
+                    this.name.buffer = new short[_name_buffers];
+                }
+                _src = _src.derive(_name_bufferi);
+                for ( int _i = 0; _i < _name_bufferl; _i++ ) {
+                    this.name.buffer[ _i ] = (short) _src.dec_ndr_short();
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 325:342, 343:360, 361:378, 770:787.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.